From 7c666ce4bbdd15f6571bd5ed4f260ec93d12300f Mon Sep 17 00:00:00 2001
From: Bing <bing.deng@intel.com>
Date: Wed, 2 Jan 2019 18:17:43 +0800
Subject: [PATCH] DO NOT MERGE According to the status of permission,
 enable/disable the Toggle of IconToggleLineItem.
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

There is no permission policy checking in AppPermissionsFragment for the automotive device.
Pop a warning dialog when fixed permission is trying to change.
Overwrite onClick() for modifying the status of IconToggleLineItem and setting the
runtime permission.

Bug: 114013494
Test: 1) Launch “com.android.cts.verifier” apk
         Start test: BYOD Managed Provisioning -> Permissions lockdown ->
         Choose “Grant” or “Deny”
      2) If built-in package installer is used, below failure are fixed:
         run cts -m CtsAppSecurityHostTestCases
             -t android.appsecurity.cts.PermissionsHostTest#testCompatRevoked22
         run cts -m CtsAppSecurityHostTestCases
             -t android.appsecurity.cts.PermissionsHostTest
                 #testGrantPreviouslyRevokedWithPrejudiceShowsPrompt23
         run cts -m CtsAppSecurityHostTestCases
             -t android.appsecurity.cts.PermissionsHostTest#testGranted23
         run cts -m CtsAppSecurityHostTestCases
             -t android.appsecurity.cts.PermissionsHostTest#testNoResidualPermissionsOnUninstall
         run cts -m CtsAppSecurityHostTestCases
             -t android.appsecurity.cts.PermissionsHostTest
                 #testRevokePropagatedOnUpgradeOldToNewModel

Change-Id: I557ddf06cdfae3d5dcfb81a51992f7e710b967ef
---
 .../permission/ui/auto/AppPermissionsFragment.java | 64 ++++++++++++++--------
 1 file changed, 42 insertions(+), 22 deletions(-)

diff --git a/src/com/android/packageinstaller/permission/ui/auto/AppPermissionsFragment.java b/src/com/android/packageinstaller/permission/ui/auto/AppPermissionsFragment.java
index 36ad469..dfd6bd7 100644
--- a/src/com/android/packageinstaller/permission/ui/auto/AppPermissionsFragment.java
+++ b/src/com/android/packageinstaller/permission/ui/auto/AppPermissionsFragment.java
@@ -40,6 +40,8 @@ import com.android.packageinstaller.R;
 import com.android.packageinstaller.permission.model.AppPermissionGroup;
 import com.android.packageinstaller.permission.model.AppPermissions;
 import com.android.packageinstaller.permission.utils.Utils;
+import com.android.settingslib.RestrictedLockUtils;
+import com.android.settingslib.RestrictedLockUtils.EnforcedAdmin;
 
 import java.util.ArrayList;
 
@@ -173,11 +175,13 @@ public final class AppPermissionsFragment extends Fragment{
     private class PermissionLineItem extends IconToggleLineItem {
         private final AppPermissionGroup mPermissionGroup;
         private final Context mContext;
+        private boolean mIsPolicyFixed = false;
 
         PermissionLineItem(AppPermissionGroup permissionGroup, Context context) {
             super(permissionGroup.getLabel(), context);
             mContext = context;
             mPermissionGroup = permissionGroup;
+            mIsPolicyFixed = permissionGroup.isPolicyFixed();
         }
 
         @Override
@@ -185,31 +189,16 @@ public final class AppPermissionsFragment extends Fragment{
             if (event.getAction() != MotionEvent.ACTION_DOWN) {
                 return true;
             }
-            if (!isChecked()) {
-                mPermissionGroup.grantRuntimePermissions(false);
-                toggleSwitch.performClick();
-            } else {
-                final boolean grantedByDefault =
-                        mPermissionGroup.hasGrantedByDefaultPermission();
-                if (grantedByDefault || !mPermissionGroup.doesSupportRuntimePermissions()) {
-                    new AlertDialog.Builder(mContext)
-                            .setMessage(grantedByDefault
-                                    ? R.string.system_warning : R.string.old_sdk_deny_warning)
-                            .setNegativeButton(R.string.cancel, null /* listener */)
-                            .setPositiveButton(R.string.grant_dialog_button_deny_anyway,
-                                    (dialog, which) -> {
-                                        mPermissionGroup.revokeRuntimePermissions(false);
-                                        toggleSwitch.performClick();
-                                    })
-                            .show();
-                } else {
-                    mPermissionGroup.revokeRuntimePermissions(false);
-                    toggleSwitch.performClick();
-                }
-            }
+
+            performClickWithItemOrToggle(toggleSwitch);
             return true;
         }
 
+        @Override
+        public void onClick(View v) {
+            performClickWithItemOrToggle(getSwitch());
+        }
+
         @DrawableRes
         public int getIcon() {
             return mPermissionGroup.getIconResId();
@@ -234,5 +223,36 @@ public final class AppPermissionsFragment extends Fragment{
         public boolean isExpandable() {
             return false;
         }
+
+        public void performClickWithItemOrToggle(Switch toggleSwitch) {
+            if(mIsPolicyFixed){
+                EnforcedAdmin enforcedAdmin = RestrictedLockUtils.getProfileOrDeviceOwner
+                        (mContext, mPermissionGroup.getUserId());
+                RestrictedLockUtils.sendShowAdminSupportDetailsIntent(mContext, enforcedAdmin);
+                return;
+            }
+            if (!isChecked()) {
+                mPermissionGroup.grantRuntimePermissions(/* fixedByTheUser= */ false);
+                toggleSwitch.performClick();
+            } else {
+                final boolean grantedByDefault =
+                        mPermissionGroup.hasGrantedByDefaultPermission();
+                if (grantedByDefault || !mPermissionGroup.doesSupportRuntimePermissions()) {
+                    new AlertDialog.Builder(mContext)
+                            .setMessage(grantedByDefault
+                                    ? R.string.system_warning : R.string.old_sdk_deny_warning)
+                            .setNegativeButton(R.string.cancel, null /* listener */)
+                            .setPositiveButton(R.string.grant_dialog_button_deny_anyway,
+                                    (dialog, which) -> {
+                                        mPermissionGroup.revokeRuntimePermissions(false);
+                                        toggleSwitch.performClick();
+                                    })
+                            .show();
+                } else {
+                    mPermissionGroup.revokeRuntimePermissions(false);
+                    toggleSwitch.performClick();
+                }
+            }
+        }
     }
 }
-- 
1.9.1

